python-oauth-web-app/app.py

```diff
@@ -0,0 +1,300 @@
+    conn = get_db()
+    cur = conn.cursor()
+    cur.execute(query, args)
+    rv = cur.fetchall()
+    conn.commit()
+    cur.close()
+    return (rv[0] if rv else None) if one else rv
+
+
+def init_db():
+    print("initializing database...")
+    with app.app_context():
+        db = get_db()
+        with app.open_resource("schema.sql", mode="r") as f:
+            db.cursor().executescript(f.read())
+        db.commit()
+
+
+init_db()
+
+
+@app.teardown_appcontext
+def close_connection(exception):
+    db = getattr(g, "_database", None)
+    if db is not None:
+        db.close()
+
+
+@app.before_request
+def load_logged_in_user():
+    user_did = session.get("user_did")
+
+    if user_did is None:
+        g.user = None
+    else:
+        g.user = (
+            get_db()
+            .execute("SELECT * FROM oauth_session WHERE did = ?", (user_did,))
+            .fetchone()
+        )
+
+
+def login_required(view):
+    @functools.wraps(view)
+    def wrapped_view(**kwargs):
+        if g.user is None:
+            return redirect("/oauth/login")
+
+        return view(**kwargs)
+
+    return wrapped_view
+
+
+@app.errorhandler(500)
+def internal_server_error(e):
+    return render_template("error.html", status_code=500, err=e), 500
+
+
+@app.errorhandler(400)
+def bad_request_error(e):
+    return render_template("error.html", status_code=400, err=e), 400
+
+
+@app.route("/")
+def homepage():
+    return render_template("home.html")
+
+
+@app.route("/oauth/client-metadata.json")
+def oauth_client_metadata():
+    app_url = request.url_root.replace("http://", "https://")
+    client_id = f"{app_url}oauth/client-metadata.json"
+
+    return jsonify(
+        {
+            # simply using the full request URL for the client_id
+            "client_id": client_id,
+            "dpop_bound_access_tokens": True,
+            "application_type": "web",
+            "redirect_uris": [f"{app_url}oauth/authorize"],
+            "client_uri": app_url,
+            "subject_type": "public",
+            "grant_types": ["authorization_code", "refresh_token"],
+            "response_types": ["code"],  # TODO: "code id_token"?
+            "scope": "openid profile offline_access",
+            "client_name": "atproto OAuth Flask Backend Demo",
+            "token_endpoint_auth_method": "none",
+            "jwks_uri": f"{app_url}oauth/jwks.json",
+            # "jwks": { #    "keys": [pub_jwk], #},
+        }
+    )
+
+
+@app.route("/oauth/jwks.json")
+def oauth_jwks():
+    return jsonify(
+        {
+            "keys": [pub_jwk],
+        }
+    )
+
+
+@app.route("/oauth/login", methods=("GET", "POST"))
+def oauth_login():
+    if request.method != "POST":
+        return render_template("login.html")
+
+    # resolve handle and DID document, using atproto identity helpers
+    username = request.form["username"]
+    if not valid_handle(username):
+        flash("Invalid Handle")
+        return render_template("login.html"), 400
```

Comment:
Note that the spec states:

> First step (1) is for the client to ask the user for their [ATPROTO](https://atproto.com/) @handle, or, should they have forgotten their handle, their PDS or Entryway's URL (e.g. bsky.social).

The TS implementation also supports a `did` as input here.

I might be worth reflecting that here.

Reply:
Will definitely add "handle or DID". I think i'll probably skip the "PDS or auth server" parts, but we can discuss those in the docs/specs.

Reply:
The "PDS or auth server" is important for people that do not have an account yet. You could "hard code" `bsky.app` as account provider, and not allow the user to pick one of their choice, but I think this flow should be implemented as an example on how to handle new users in 3rd party apps.

Reply:
added support for this